website/newsletters/forms.py

```diff
@@ -43,3 +33,125 @@ class Media:
+
+    remove_existing = forms.BooleanField(required=False, initial=False)
+
+    sort_by = forms.ChoiceField(
+        choices=(
+            ("start", _("Event start date")),
+            ("registration", _("Registrations open date")),
+        ),
+        required=True,
+    )
+
+    def import_events(self, newsletter: Newsletter):
+        """Import events from the database into the newsletter."""
+        if self.cleaned_data["remove_existing"]:
+            NewsletterEvent.objects.filter(newsletter=newsletter).delete()
+
+        events = Event.objects.filter(published=True)
+
+        if self.cleaned_data["event_start"] and self.cleaned_data["registration_start"]:
+            events = events.filter(
+                start__date__lte=self.cleaned_data["event_end"],
+                start__date__gte=self.cleaned_data["event_start"],
+                registration_start__date__gte=self.cleaned_data["registration_start"],
+                registration_start__date__lte=self.cleaned_data["registration_end"],
+            )
+        elif self.cleaned_data["event_start"]:
+            events = events.filter(
+                start__date__gte=self.cleaned_data["event_start"],
+                start__date__lte=self.cleaned_data["event_end"],
+            )
+        elif self.cleaned_data["registration_start"]:
+            events = events.filter(
+                registration_start__date__gte=self.cleaned_data["registration_start"],
+                registration_start__date__lte=self.cleaned_data["registration_end"],
+            )
+
+        if self.cleaned_data["sort_by"] == "start":
+            events = events.order_by("start")
+        elif self.cleaned_data["sort_by"] == "registration":
+            events = events.order_by("registration_start")
+
+        max_order = (
+            NewsletterEvent.objects.filter(newsletter=newsletter)
+            .order_by("-order")
+            .first()
+        )
+        max_order = max_order.order if max_order else 0
+
+        for i, event in enumerate(events, start=max_order + 1):
+            NewsletterEvent.objects.create(
+                newsletter=newsletter,
+                order=i,
+                title=event.title,
+                url=settings.BASE_URL + event.get_absolute_url(),
+                description=event.description,
+                where=event.location,
+                start_datetime=event.start,
+                end_datetime=event.end,
+                show_costs_warning=event.fine > 0,
+                price=event.price,
+                penalty_costs=event.fine,
+            )
+
+    def clean(self):
+        cleaned_data = super().clean()
+        if (
+            cleaned_data.get("event_start") is not None
+            and cleaned_data.get("event_end") is None
+        ):
+            self.add_error(
+                "event_end", _("This field is required if you specify a start date.")
+            )
+        if (
```

Comment:
I feel like there's a way to make this logic less ugly? Especially since Black formatted it in a horrible way.